### PR TITLE
add CRUD for organization information

### PR DESCRIPTION
**Pull request by LukaszPiskadlo**

### CODE REVIEW

File: src/main/java/com/intive/patronage/toz/exception/AlreadyExistsException.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.intive.patronage.toz.exception;
+
+public class AlreadyExistsException extends RuntimeException {
+    private final static String ALREADY_EXISTS = " already exists!";
+
+    public AlreadyExistsException(String name) {
+        super("'" + name + "'" + ALREADY_EXISTS);
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/exception/AlreadyExistsException.java]: Use `String.format()` instead.

File: src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.validation.constraints.Pattern;
+
+@Entity
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class OrganizationInfo {
+
+    @Id
+    private final Long id;
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: Let's use `UUID` instead.

File: src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    private final String phone;
+    private final String fax;
+    @URL
+    private final String website;
+
+    @NotEmpty
+    private final String accountNumber;
+    private final String bankName;
+
+    OrganizationInfo() {
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: Could we live without `OrganiaztionInfo` which contains null's? :) Could you remove this constructor?

File: src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        return city;
+    }
+
+    public String getStreet() {
+        return street;
+    }
+
+    @JsonIgnore
+    public String getFullAddress() {
+        return street + " " + postCode + " " + city;
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: Use `String.format()` instead.

File: src/main/java/com/intive/patronage/toz/service/OrganizationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import com.intive.patronage.toz.exception.AlreadyExistsException;
+import com.intive.patronage.toz.exception.ArgumentNotValidException;
+import com.intive.patronage.toz.exception.NotFoundException;
+import com.intive.patronage.toz.model.OrganizationInfo;
+import com.intive.patronage.toz.repository.OrganizationInfoRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class OrganizationServiceImpl implements OrganizationService {
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/service/OrganizationServiceImpl.java]: I hate `*Impl` names which are not informative. Rename to `SingleOrganizationService` or `DefaultOrganizationService`.

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 6]
```diff
@@ -1,27 +1,81 @@
 databaseChangeLog:
   - changeSet:
       id: 1
-      author: mlotysz
+      author: lukaszpiskadlo
```
Comment (mlotysz) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: Note only: we haven't got DB yet and thus change DB schema is OK. But in the future you'll need to create new change set and create migration script for data too.

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 7]
```diff
@@ -1,27 +1,81 @@
...
+                  name: phone
+                  type: varchar(64)
+              - column:
+                  name: fax
+                  type: varchar(64)
+              - column:
+                  name: website
+                  type: varchar(255)
+              - column:
+                  name: account_number
```
Comment (mlotysz) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: I think, `bank_account_number` wont confuse team.

File: src/test/java/sample/liquibase/LiquibaseApplicationTests.java [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
 import org.junit.Test;
-
 import org.springframework.boot.test.rule.OutputCapture;
 import org.springframework.core.NestedCheckedException;
 
+import java.net.ConnectException;
+
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class LiquibaseApplicationTests {
```
Comment (mlotysz) [inline src/test/java/sample/liquibase/LiquibaseApplicationTests.java]: Just remove this redundant test?

File: src/main/java/com/intive/patronage/toz/exception/AlreadyExistsException.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.intive.patronage.toz.exception;
+
+public class AlreadyExistsException extends RuntimeException {
+    private final static String ALREADY_EXISTS = " already exists!";
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/exception/AlreadyExistsException.java]: Put all error messages to message bundle.

File: src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java (hunk 3 above)
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: JPA requires default constructor, which cant be empty if fields are final.

File: src/main/java/com/intive/patronage/toz/controller/OrganizationController.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping(value = "/organization", produces = MediaType.APPLICATION_JSON_VALUE)
+public class OrganizationController {
```
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/controller/OrganizationController.java]: Add basic swagger documentation.

File: src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java (hunk 3 above)
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: I see. You mix JPA and frontend view in same class. Do you think you can separate data model (JPA) and external model (JSON) somehow? I think those two should live with different packages. Let me check google...

Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/model/OrganizationInfo.java]: Found:
- http://stackoverflow.com/questions/35238466/spring-boot-decision-to-create-dto-object-separately-both-for-rest-and-jpa and http://www.cherryshoetech.com/2016/05/jpa-json-serializationdeserialization.html

Now I cannot prove that my option is the way we should go. Let's leave it for now as is. I bet, we return to this discussion in the future. :)

**Comment by macieg-b** [comment]
It's ok

File: src/main/java/com/intive/patronage/toz/controller/OrganizationController.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    private final OrganizationService service;
+
+    @Autowired
+    public OrganizationController(OrganizationService service) {
+        this.service = service;
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/controller/OrganizationController.java]: Can it be simplyfied to:
```java
@Autowired
private final OrganizationService service;
```

File: src/main/java/com/intive/patronage/toz/exception/ArgumentNotValidException.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.intive.patronage.toz.exception;
+
+public class ArgumentNotValidException extends RuntimeException {
```
Comment (pskrzypek84) [inline src/main/java/com/intive/patronage/toz/exception/ArgumentNotValidException.java]: In service you throws this exception if client pass organization `ID`, but you do not map this to proper error code.

**Comment by mlotysz** [comment]
Looks a way better! I'm gonna merge it. Thanks!

**Pull request merged**